src/v/kafka/protocol/batch_reader.cc

```diff
@@ -116,4 +116,26 @@ kafka_batch_adapter batch_reader::consume_batch() {
     return kba;
 }
 
+ss::future<batch_reader::storage_t>
+batch_reader::do_load_slice(model::timeout_clock::time_point tp) {
+    using data_t = model::record_batch_reader::data_t;
+    return ss::do_with(data_t{}, [this, tp](data_t& batches) {
+        const auto resources_exceeded = [this, tp] {
+            return is_end_of_stream() || model::timeout_clock::now() > tp;
+        };
+        const auto consume_one = [this, &batches]() {
+            auto kba = consume_batch();
+            if (likely(kba.v2_format && kba.valid_crc && kba.batch)) {
+                batches.push_back(std::move(*kba.batch));
+            } else {
+                _do_load_slice_failed = true;
```

Comment:
I was tempted, and then I can drop the `_do_load_slice_failed` member, WDYT?

Reply:
I kinda like `_do_load_slice_failed` because it makes the termination condition clear. Would it make sense to clear batches and immediately return an exceptional future? I'm not sure how you were planning on communicating an error to a consumer.

Reply:
I thought about it as a better signalling mechanism, but a quick search through the codebase didn't reveal any other readers that do that, and I wasn't sure what was idiomatic.

Reply:
I think log reader is the only record batch reader that has an opportunity to fail. Here is where it returns an exceptional future from `do_load_slice`

https://github.com/vectorizedio/redpanda/blob/dev/src/v/storage/log_reader.cc#L297